File: leetcode/字符串匹配/0139.py
# 139. 单词拆分
# https://leetcode-cn.com/problems/word-break/

# https://leetcode-cn.com/problems/word-break/solution/dan-ci-chai-fen-by-leetcode-solution/
# https://leetcode-cn.com/problems/word-break/solution/shou-hui-tu-jie-san-chong-fang-fa-dfs-bfs-dong-tai/

# "leetcode"
# ["leet", "code"]
# "acaaaaabbbdbcccdcdaadcdccacbcccabbbbcdaaaaaadb"
# ["abbcbda", "cbdaaa", "b", "dadaaad", "dccbbbc", "dccadd", "ccbdbc", "bbca", "bacbcdd", "a", "bacb", "cbc", "adc", "c", "cbdbcad", "cdbab", "db", "abbcdbd", "bcb", "bbdab", "aa", "bcadb", "bacbcb", "ca", "dbdabdb",
#     "ccd", "acbb", "bdc", "acbccd", "d", "cccdcda", "dcbd", "cbccacd", "ac", "cca", "aaddc", "dccac", "ccdc", "bbbbcda", "ba", "adbcadb", "dca", "abd", "bdbb", "ddadbad", "badb", "ab", "aaaaa", "acba", "abbb"]

# 2. 官方解 - 动态规划


# 3. BFS + 避免遍历重复节点


# 2. 递归 + 记忆化


# 1. 自解 - 笨方法 字典 + 内部递归
# 执行结果：超时
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dct = {}
        for word in wordDict:
            dct[word] = True
        # print(dct)

        def check(v: str) -> bool:
            # print(str(["check v", v]))
            n = len(v)
            for i in range(n):
                s0, s1 = v[0:i], v[i:]
                pre = check(s0)
                # print(str([s0, s1, s1 in dct, pre]))
                if s1 in dct and (i <= 0 or pre):
                    return True
            return False

        return check(s)

File: leetcode/字符串匹配/0139.go
// 139. 单词拆分
// https://leetcode-cn.com/problems/word-break/

// https://leetcode-cn.com/problems/word-break/solution/dan-ci-chai-fen-by-leetcode-solution/
// https://leetcode-cn.com/problems/word-break/solution/shou-hui-tu-jie-san-chong-fang-fa-dfs-bfs-dong-tai/

// "leetcode"
// ["leet", "code"]
// "acaaaaabbbdbcccdcdaadcdccacbcccabbbbcdaaaaaadb"
// ["abbcbda", "cbdaaa", "b", "dadaaad", "dccbbbc", "dccadd", "ccbdbc", "bbca", "bacbcdd", "a", "bacb", "cbc", "adc", "c", "cbdbcad", "cdbab", "db", "abbcdbd", "bcb", "bbdab", "aa", "bcadb", "bacbcb", "ca", "dbdabdb",
//     "ccd", "acbb", "bdc", "acbccd", "d", "cccdcda", "dcbd", "cbccacd", "ac", "cca", "aaddc", "dccac", "ccdc", "bbbbcda", "ba", "adbcadb", "dca", "abd", "bdbb", "ddadbad", "badb", "ab", "aaaaa", "acba", "abbb"]

// 官方解 - 动态规划
// 执行用时： 0 ms , 在所有 Go 提交中击败了 100.00% 的用户
// 内存消耗： 2.2 MB , 在所有 Go 提交中击败了 100.00% 的用户
func wordBreak(s string, wordDict []string) bool {
    wordDictSet := make(map[string]bool)
    for _, w := range wordDict {
        wordDictSet[w] = true
    }
    dp := make([]bool, len(s) + 1)
    dp[0] = true
    for i := 1; i <= len(s); i++ {
        for j := 0; j < i; j++ {
            if dp[j] && wordDictSet[s[j:i]] {
                dp[i] = true
                break
            }
        }
    }
    return dp[len(s)]
}
